i push an edit of commit message because test fail on something which is not related to my dev:
Caused by: java.lang.NullPointerException
	at org.jboss.jandex.Indexer.decodeClassEntry(Indexer.java:1298)

--- a/core/runtime/src/main/java/io/quarkus/runtime/TemplateHtmlBuilder.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/TemplateHtmlBuilder.java
@@ -25,6 +25,8 @@ public class TemplateHtmlBuilder {
 
     private static final String RESOURCES_START = "<div class=\"intro\">%1$s</div><div class=\"resources\">";
 
+    private static final String ANCHOR_TEMPLATE = "<a href=\"/%1$s\">/%2$s</a>";
+
     private static final String RESOURCE_TEMPLATE = "<h3>%1$s</h3>\n";
 
     private static final String LIST_START = "<ul>\n";
@@ -36,8 +38,6 @@ public class TemplateHtmlBuilder {
 
     private static final String METHOD_END = "    </ul>\n"
             + "</li>";
-
-    private static final String SERVLET_MAPPING = "<li>%1$s</li>\n";
 
     private static final String LIST_END = "</ul>\n";
 
@@ -166,8 +166,28 @@ public class TemplateHtmlBuilder {
     }
 
     public TemplateHtmlBuilder resourcePath(String title) {
-        result.append(String.format(RESOURCE_TEMPLATE, escapeHtml(title)));
-        result.append(LIST_START);
+        return resourcePath(title, true, false);
+    }
+
+    public TemplateHtmlBuilder staticResourcePath(String title) {
+        return resourcePath(title, false, true);
+    }
+
+    public TemplateHtmlBuilder servletMapping(String title) {
+        return resourcePath(title, false, false);
+    }
+
+    private TemplateHtmlBuilder resourcePath(String title, boolean withListStart, boolean withAnchor) {
+        String content;
+        if (withAnchor) {
+            content = String.format(ANCHOR_TEMPLATE, title, escapeHtml(title));
+        } else {
+            content = escapeHtml(title);
+        }
+        result.append(String.format(RESOURCE_TEMPLATE, content));
+        if (withListStart) {
+            result.append(LIST_START);
+        }
         return this;
     }
 
@@ -216,10 +236,4 @@ public class TemplateHtmlBuilder {
                 .replace("<", "&lt;")
                 .replace(">", "&gt;");
     }
-
-    public TemplateHtmlBuilder servletMapping(String servletMapping) {
-        result.append(String.format(SERVLET_MAPPING, servletMapping));
-        return this;
-    }
-
 }

--- a/core/runtime/src/main/java/io/quarkus/runtime/TemplateHtmlBuilder.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/TemplateHtmlBuilder.java
@@ -23,7 +23,7 @@ public class TemplateHtmlBuilder {
             "</header>\n" +
             "<div class=\"container content\">\n";
 
-    private static final String RESOURCES_START = "<div class=\"intro\">If you are looking for your REST resources, here are the ones we discovered:</div><div class=\"resources\">";
+    private static final String RESOURCES_START = "<div class=\"intro\">%1$s</div><div class=\"resources\">";
 
     private static final String RESOURCE_TEMPLATE = "<h3>%1$s</h3>\n";
 
@@ -36,6 +36,8 @@ public class TemplateHtmlBuilder {
 
     private static final String METHOD_END = "    </ul>\n"
             + "</li>";
+
+    private static final String SERVLET_MAPPING = "<li>%1$s</li>\n";
 
     private static final String LIST_END = "</ul>\n";
 
@@ -148,8 +150,8 @@ public class TemplateHtmlBuilder {
         return this;
     }
 
-    public TemplateHtmlBuilder resourcesStart() {
-        result.append(RESOURCES_START);
+    public TemplateHtmlBuilder resourcesStart(String title) {
+        result.append(String.format(RESOURCES_START, title));
         return this;
     }
 
@@ -189,6 +191,11 @@ public class TemplateHtmlBuilder {
         return this;
     }
 
+    public TemplateHtmlBuilder resourceStart() {
+        result.append(LIST_START);
+        return this;
+    }
+
     public TemplateHtmlBuilder resourceEnd() {
         result.append(LIST_END);
         return this;
@@ -209,4 +216,10 @@ public class TemplateHtmlBuilder {
                 .replace("<", "&lt;")
                 .replace(">", "&gt;");
     }
+
+    public TemplateHtmlBuilder servletMapping(String servletMapping) {
+        result.append(String.format(SERVLET_MAPPING, servletMapping));
+        return this;
+    }
+
 }

--- a/extensions/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/ResteasyProcessor.java
+++ b/extensions/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/ResteasyProcessor.java
@@ -6,6 +6,7 @@ import static io.quarkus.deployment.annotations.ExecutionTime.STATIC_INIT;
 import java.util.List;
 import java.util.Map.Entry;
 import java.util.Optional;
+import java.util.Set;
 import java.util.stream.Collectors;
 
 import javax.servlet.DispatcherType;
@@ -28,6 +29,7 @@ import io.quarkus.resteasy.server.common.deployment.ResteasyServerConfigBuildIte
 import io.quarkus.undertow.deployment.FilterBuildItem;
 import io.quarkus.undertow.deployment.ServletBuildItem;
 import io.quarkus.undertow.deployment.ServletInitParamBuildItem;
+import io.quarkus.undertow.deployment.StaticResourceFilesBuildItem;
 
 /**
  * Processor that finds JAX-RS classes in the deployment
@@ -95,9 +97,24 @@ public class ResteasyProcessor {
 
     @BuildStep(onlyIf = IsDevelopment.class)
     @Record(STATIC_INIT)
-    void setupExceptionMapper(List<ServletBuildItem> servlets) {
+    void addServletsToExceptionMapper(List<ServletBuildItem> servlets) {
         NotFoundExceptionMapper.servlets(servlets.stream().filter(s -> !JAX_RS_SERVLET_NAME.equals(s.getName()))
                 .collect(Collectors.toMap(s -> s.getName(), s -> s.getMappings())));
+    }
+
+    @BuildStep(onlyIf = IsDevelopment.class)
+    @Record(STATIC_INIT)
+    void addStaticResourcesExceptionMapper(StaticResourceFilesBuildItem paths) {
+        //limit to 1000 to not have to many files to display
+        Set<String> staticResources = paths.files.stream().filter(this::isHtmlFileName).limit(1000).collect(Collectors.toSet());
+        if (staticResources.isEmpty()) {
+            staticResources = paths.files.stream().limit(1000).collect(Collectors.toSet());
+        }
+        NotFoundExceptionMapper.staticResources(staticResources);
+    }
+
+    private boolean isHtmlFileName(String fileName) {
+        return fileName.endsWith(".html") || fileName.endsWith(".htm");
     }
 
     private String getMappingPath(String path) {

--- a/extensions/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/ResteasyProcessor.java
+++ b/extensions/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/ResteasyProcessor.java
@@ -1,9 +1,12 @@
 package io.quarkus.resteasy.deployment;
 
 import static io.quarkus.deployment.annotations.ExecutionTime.RUNTIME_INIT;
+import static io.quarkus.deployment.annotations.ExecutionTime.STATIC_INIT;
 
+import java.util.List;
 import java.util.Map.Entry;
 import java.util.Optional;
+import java.util.stream.Collectors;
 
 import javax.servlet.DispatcherType;
 import javax.ws.rs.core.Application;
@@ -88,6 +91,13 @@ public class ResteasyProcessor {
     @Record(RUNTIME_INIT)
     void setupExceptionMapper(BuildProducer<ResteasyJaxrsProviderBuildItem> providers) {
         providers.produce(new ResteasyJaxrsProviderBuildItem(NotFoundExceptionMapper.class.getName()));
+    }
+
+    @BuildStep(onlyIf = IsDevelopment.class)
+    @Record(STATIC_INIT)
+    void setupExceptionMapper(List<ServletBuildItem> servlets) {
+        NotFoundExceptionMapper.servlets(servlets.stream().filter(s -> !JAX_RS_SERVLET_NAME.equals(s.getName()))
+                .collect(Collectors.toMap(s -> s.getName(), s -> s.getMappings())));
     }
 
     private String getMappingPath(String path) {

--- a/extensions/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/NotFoundExceptionMapper.java
+++ b/extensions/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/NotFoundExceptionMapper.java
@@ -10,8 +10,8 @@ import java.util.HashMap;
 import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
-import java.util.Map.Entry;
 import java.util.Set;
+import java.util.stream.Collectors;
 
 import javax.annotation.Priority;
 import javax.ws.rs.NotFoundException;
@@ -42,7 +42,8 @@ public class NotFoundExceptionMapper implements ExceptionMapper<NotFoundExceptio
     private final static Variant HTML_VARIANT = new Variant(MediaType.TEXT_HTML_TYPE, (String) null, null);
     private final static List<Variant> VARIANTS = Arrays.asList(JSON_VARIANT, HTML_VARIANT);
 
-    private static Map<String, List<String>> servletToMapping;
+    private static List<String> servletMappings;
+    private static List<String> staticResources;
 
     @Context
     private Registry registry = null;
@@ -63,12 +64,11 @@ public class NotFoundExceptionMapper implements ExceptionMapper<NotFoundExceptio
             this.produces = produces;
             this.consumes = consumes;
         }
-
     }
 
     public static final class ResourceDescription {
-        public String basePath;
-        public List<MethodDescription> calls;
+        public final String basePath;
+        public final List<MethodDescription> calls;
 
         public ResourceDescription(String basePath) {
             this.basePath = basePath;
@@ -156,6 +156,7 @@ public class NotFoundExceptionMapper implements ExceptionMapper<NotFoundExceptio
         List<ResourceDescription> descriptions = ResourceDescription
                 .fromBoundResourceInvokers(bounded
                         .entrySet());
+
         return respond(descriptions);
     }
 
@@ -203,14 +204,18 @@ public class NotFoundExceptionMapper implements ExceptionMapper<NotFoundExceptio
             }
             sb.resourcesEnd();
 
-            if (!servletToMapping.isEmpty()) {
+            if (!servletMappings.isEmpty()) {
                 sb.resourcesStart("Servlet mappings");
-                for (Entry<String, List<String>> servletMappings : servletToMapping.entrySet()) {
-                    sb.resourcePath(servletMappings.getKey());
-                    for (String mapping : servletMappings.getValue()) {
-                        sb.servletMapping(mapping);
-                    }
-                    sb.resourceEnd();
+                for (String servletMapping : servletMappings) {
+                    sb.servletMapping(servletMapping);
+                }
+                sb.resourcesEnd();
+            }
+
+            if (!staticResources.isEmpty()) {
+                sb.resourcesStart("Static resources");
+                for (String staticResource : staticResources) {
+                    sb.staticResourcePath(staticResource);
                 }
                 sb.resourcesEnd();
             }
@@ -227,6 +232,13 @@ public class NotFoundExceptionMapper implements ExceptionMapper<NotFoundExceptio
     }
 
     public static void servlets(Map<String, List<String>> servletToMapping) {
-        NotFoundExceptionMapper.servletToMapping = servletToMapping;
+        NotFoundExceptionMapper.servletMappings = servletToMapping.values().stream()
+                .flatMap(List::stream)
+                .sorted()
+                .collect(Collectors.toList());
+    }
+
+    public static void staticResources(Set<String> knownFiles) {
+        NotFoundExceptionMapper.staticResources = knownFiles.stream().sorted().collect(Collectors.toList());
     }
 }

--- a/extensions/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/NotFoundExceptionMapper.java
+++ b/extensions/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/NotFoundExceptionMapper.java
@@ -10,6 +10,7 @@ import java.util.HashMap;
 import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
+import java.util.Map.Entry;
 import java.util.Set;
 
 import javax.annotation.Priority;
@@ -40,6 +41,8 @@ public class NotFoundExceptionMapper implements ExceptionMapper<NotFoundExceptio
     private final static Variant JSON_VARIANT = new Variant(MediaType.APPLICATION_JSON_TYPE, (String) null, null);
     private final static Variant HTML_VARIANT = new Variant(MediaType.TEXT_HTML_TYPE, (String) null, null);
     private final static List<Variant> VARIANTS = Arrays.asList(JSON_VARIANT, HTML_VARIANT);
+
+    private static Map<String, List<String>> servletToMapping;
 
     @Context
     private Registry registry = null;
@@ -179,8 +182,8 @@ public class NotFoundExceptionMapper implements ExceptionMapper<NotFoundExceptio
         }
 
         if (variant == HTML_VARIANT) {
-            TemplateHtmlBuilder sb = new TemplateHtmlBuilder("404 - Resource Not Found", "", "REST interface overview");
-            sb.resourcesStart();
+            TemplateHtmlBuilder sb = new TemplateHtmlBuilder("404 - Resource Not Found", "", "Resources overview");
+            sb.resourcesStart("REST resources");
             for (ResourceDescription resource : descriptions) {
                 sb.resourcePath(resource.basePath);
                 for (MethodDescription method : resource.calls) {
@@ -200,9 +203,20 @@ public class NotFoundExceptionMapper implements ExceptionMapper<NotFoundExceptio
             }
             sb.resourcesEnd();
 
+            if (!servletToMapping.isEmpty()) {
+                sb.resourcesStart("Servlet mappings");
+                for (Entry<String, List<String>> servletMappings : servletToMapping.entrySet()) {
+                    sb.resourcePath(servletMappings.getKey());
+                    for (String mapping : servletMappings.getValue()) {
+                        sb.servletMapping(mapping);
+                    }
+                    sb.resourceEnd();
+                }
+                sb.resourcesEnd();
+            }
+
             return Response.status(Status.NOT_FOUND).entity(sb.toString()).type(MediaType.TEXT_HTML_TYPE).build();
         }
-
         return Response.status(Status.NOT_FOUND).build();
     }
 
@@ -210,5 +224,9 @@ public class NotFoundExceptionMapper implements ExceptionMapper<NotFoundExceptio
         ServerDrivenNegotiation negotiation = new ServerDrivenNegotiation();
         negotiation.setAcceptHeaders(headers.getRequestHeaders().get(ACCEPT));
         return negotiation.getBestMatch(VARIANTS);
+    }
+
+    public static void servlets(Map<String, List<String>> servletToMapping) {
+        NotFoundExceptionMapper.servletToMapping = servletToMapping;
     }
 }

--- a/extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/StaticResourceFilesBuildItem.java
+++ b/extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/StaticResourceFilesBuildItem.java
@@ -1,0 +1,14 @@
+package io.quarkus.undertow.deployment;
+
+import java.util.Set;
+
+import io.quarkus.builder.item.SimpleBuildItem;
+
+public final class StaticResourceFilesBuildItem extends SimpleBuildItem {
+
+    public final Set<String> files;
+
+    StaticResourceFilesBuildItem(Set<String> files) {
+        this.files = files;
+    }
+}

--- a/extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowStaticResourcesBuildStep.java
+++ b/extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowStaticResourcesBuildStep.java
@@ -36,15 +36,13 @@ public class UndertowStaticResourcesBuildStep {
     }
 
     @BuildStep
-    KnownPathsBuildItem scanStaticResources(ApplicationArchivesBuildItem applicationArchivesBuildItem,
+    void scanStaticResources(ApplicationArchivesBuildItem applicationArchivesBuildItem,
             BuildProducer<GeneratedResourceBuildItem> generatedResources,
+            BuildProducer<KnownPathsBuildItem> knownPathsBuilds,
+            BuildProducer<StaticResourceFilesBuildItem> staticResourceFiles,
             List<GeneratedWebResourceBuildItem> generatedWebResources,
             LaunchModeBuildItem launchModeBuildItem) throws Exception {
-        if (launchModeBuildItem.getLaunchMode() == LaunchMode.DEVELOPMENT) {
-            //we don't need to do this in development mode
-            //we serve directly from the project dir
-            return new KnownPathsBuildItem(Collections.emptySet(), Collections.emptySet());
-        }
+
         //we need to check for web resources in order to get welcome files to work
         //this kinda sucks
         Set<String> knownFiles = new HashSet<>();
@@ -105,7 +103,15 @@ public class UndertowStaticResourcesBuildStep {
                 }
             }
         }
-        return new KnownPathsBuildItem(knownFiles, knownDirectories);
+        if (launchModeBuildItem.getLaunchMode() == LaunchMode.DEVELOPMENT) {
+            //we don't need knownPaths in development mode
+            //we serve directly from the project dir
+            knownPathsBuilds.produce(new KnownPathsBuildItem(Collections.emptySet(), Collections.emptySet()));
+            //but we need files to display them in case of 404 in development mode
+            staticResourceFiles.produce(new StaticResourceFilesBuildItem(knownFiles));
+        } else {
+            knownPathsBuilds.produce(new KnownPathsBuildItem(knownFiles, knownDirectories));
+        }
     }
 
     @BuildStep

